### PR TITLE
[DB-1803] Fix projection stall race condition when writing empty transactions

### PR DESCRIPTION
**Pull request by timothycoleman**
### User Description

If a projection had finished processing all events in its input stream, and then another event was written to its input stream, under certain conditions a race condition meant that the projection would remain idle and not process that new event.

The projection would process that new event (and any subsequent events) if any event was written to any stream. i.e. the stalled projection could be recovered by writing any event to any stream.

The stall requires transactions without any events in to be written.

When a projection reaches the end of its input stream, it subscribes to the AwakerService to be notified when there are more events to process.

When a new event is committed, AwakerService decides which subscribers need to be notified. However, it only notifies them when the indexing process reaches the end of the log or a threshold is reached. (This prevents unnecessary notifications and resubscriptions).

Unfortunately, the AwakerService was only notified when the indexing process reaches the end of the log if the last record in the log was an event. This was almost always the case, but is not the case if the last record is an empty transaction containing no events (i.e. the client sent an append request with 0 events in).

A stall could therefore occur if a projection has subscribed to the AwakerService, then an event E that it is interested in is written, immediately followed by an empty transaction before E is indexed (this is the race). The AwakerService would then know that the projection needs notifying, but will not notify it, waiting instead to be notified that the indexing process has reached the end of the log, which it never will, unless another event is written.

This PR fixes the race condition by having the indexing process send a notification when the end of the log is reached, even if the last record is not an event.

___

### **Auto-created Ticket**

[#5429](https://github.com/kurrent-io/KurrentDB/issues/5429)

### **PR Type**
Bug fix

___

### **Description**
- Rename `TfEofAtNonCommitRecord` to `IndexedToEndOfTransactionFile` for clarity

- Publish `IndexedToEndOfTransactionFile` when empty transactions reach end of log

- Add `transactionPosition` parameter to `AddPendingPrepare` method signature

- Remove unused `_commitsAfterEof` flag and related EOF detection logic

- Add comprehensive test for empty transaction handling at log end

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Empty Transaction<br/>at Log End"] -->|"AddPendingPrepare<br/>with transactionPosition"| B["IndexCommitterService"]
  B -->|"ProcessCommitReplicated<br/>detects empty + EOF"| C["Publish<br/>IndexedToEndOfTransactionFile"]
  C -->|"Handle message"| D["AwakeService<br/>notifies subscribers"]
  E["StorageChaser"] -->|"CommitPendingTransaction<br/>passes transactionPosition"| B
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>StorageMessage.cs</strong><dd><code>Rename message class and add documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-3bf916b220b3b81b942409224dcc55db92205b0eca2725e76b87dfa1ae4af6ba">+5/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>AwakeService.cs</strong><dd><code>Update message handler for renamed message type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-d83573a92e17d1fae7bde0af9f13f8eb1dfc313cecef723ba78320e67ae2fb30">+5/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>IndexCommitter.cs</strong><dd><code>Update published message to renamed type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-6089e24c5822f93a140495917b273c8bf9603abb477676810f3a84fdcc13b819">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>TestFixtureWithExistingEvents.cs</strong><dd><code>Update published message to renamed type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-cdfd400d877b49a1773ad92d8c1a9d01ab829b4ac89923a0911f4745aaaffc95">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>ProjectionsSubsystem.cs</strong><dd><code>Update message subscription to renamed type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-4befecafca3c8b086a66ae8bcc8dae41acc17b9d5a709674b73a3d580c3f8d8e">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>IndexCommitterService.cs</strong><dd><code>Add transactionPosition parameter and empty transaction handling</code></dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-f70568a148ba85ebe5e4b6b0c3471b3a28fc98412cb99edf85f0828555d2b654">+5/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>StorageChaser.cs</strong><dd><code>Remove EOF detection logic and update method call</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-1aa3a3d56d7eeb4fc33cb4aa06bb5853ebe24b7a47433618b514f020fa2baf1b">+2/-9</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>PrepareLogRecord.cs</strong><dd><code>Clarify documentation for flags and ExpectedVersion</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-fd6753cacac0ef9479709e33df08d8862560bcaea4bdcbf67cca62b45cf77f7b">+6/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>when_index_committer_service_commits_empty_transaction_at_end_of_log.cs</strong><dd><code>Add test for empty transaction at log end</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-af438966e6d5ca3297f6b71ae47b82be09b238b2c31c56fe63ec81888b08992b">+47/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>with_index_committer_service.cs</strong><dd><code>Update test infrastructure for new message type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-e350e8ecc2a91e82edd2f7104df6e71592e8d639b6535b8ac993c07ee50bf8d2">+6/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>when_chaser_reads_commit_event.cs</strong><dd><code>Update method call with transactionPosition parameter</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-ee8220668e387ff3400d625712e867540568cefae5c336508b6515f25c543dde">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>FakeIndexCommitterService.cs</strong><dd><code>Update method signature with transactionPosition parameter</code></dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-cfbb77ba16c89dfaf94419f2415074cd29defb53e9a816c06626fbe8653e5caf">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>TestFixtureWithExistingEvents.cs</strong><dd><code>Update message subscription to renamed type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-04adb95fd01200fa2d45cba0ed994da826484f6a0567daa997219b12eee8cd11">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>specification_with_projection_management_service.cs</strong><dd><code>Update message subscription to renamed type</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/kurrent-io/KurrentDB/pull/5428/files#diff-9f6b52bd46e38fb4d0784fb40b967f917f5c980775ed83c21626633031a5e3bd">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: src/KurrentDB.Core/TransactionLog/LogRecords/PrepareLogRecord.cs [hunk 1]
```diff
@@ -59,7 +59,10 @@ public sealed class PrepareLogRecord : LogRecord, IEquatable<PrepareLogRecord>,
 	public PrepareFlags Flags { get; }
 	public long TransactionPosition { get; }
 	public int TransactionOffset { get; }
-	public long ExpectedVersion { get; } // if IsCommitted is set, this is final EventNumber
+	// if IsCommitted is set then ExpectedVersion is EventNumber - 1 else ExpectedVersion is Any (-2)
+	// we don't store the event number directly because this record does not necessarily have one.
```
Comment (shaan1337) [inline src/KurrentDB.Core/TransactionLog/LogRecords/PrepareLogRecord.cs]:
```suggestion
	// we store the expected version instead of the event number directly because this record does not necessarily have one.
```
